scabha/cargo.py

```diff
@@ -266,6 +262,37 @@ def natify(value):
 
         self._is_input = True
 
+    def _type_eval(self, namespace: Dict, type_string: str = None, maxlen=50):
+        """
+        Evaluate a string type hint safely. This uses the `eval` function,
+        so we perfom basic tests to avoid malicios strings
+        """
+        type_string = type_string or self.dtype
+        # remove functions and other callables
+        safespace = {}
+        for key, val in namespace.items():
+            if inspect.isfunction(val) or inspect.ismethod(val) or key.startswith("__"):
+                continue
+            elif isinstance(val, type):
+                safespace[key] = val
+            elif getattr(val, "__module__", False) in ["typing", "scabha.basetypes"]:
+                safespace[key] = val
+            # this condition has to come after looking for __module__
```

Comment:
As far as I can tell, ``safespace`` will always be the same (we only call ``_type_eval`` in this one single context, right?) So I suggest pre-cooking and saving it in a static class variable, rather than recomputing it each time. There's a metric shitton of dtype strings being parsed each time we load a recipe, so any cycles spared in this function are good for the user and the climate.